# Data Preparation

[Original Notebook Source from *Data Science: Introduction to Machine Learning for Data Science Python and Machine Learning Studio by Lee Stott*](https://github.com/leestott/intro-Datascience/blob/master/Course%20Materials/4-Cleaning_and_Manipulating-Reference.ipynb)

## Exploring `DataFrame` information

> **Learning goal:** By the end of this subsection, you should be comfortable finding general information about the data stored in pandas DataFrames.

Once you have loaded your data into pandas, it will more likely than not be in a `DataFrame`. However, if the dataset in your `DataFrame` has 60,000 rows and 400 columns, how do you even begin to get a sense of what you're working with? Fortunately, pandas provides some convenient tools to quickly look at overall information about a `DataFrame` in addition to the first few and last few rows.

In order to explore this functionality, we will import the Python scikit-learn library and use an iconic dataset that every data scientist has seen hundreds of times: British biologist Ronald Fisher's *Iris* data set used in his 1936 paper "The use of multiple measurements in taxonomic problems":

In [1]:
import pandas as pd
from sklearn.datasets import load_iris

iris = load_iris()
iris_df = pd.DataFrame(data=iris["data"], columns=iris["feature_names"])

### `DataFrame.shape`

We have loaded the Iris Dataset in the variable `iris_df`. Before diving into the data, it would be valuable to know the number of datapoints we have and the overall size of the dataset. It is useful to look at the volume of data we are dealing with.

In [2]:
iris_df.shape

(150, 4)

So, we are dealing with 150 rows and 4 columns of data. Each row represents one datapoint and each column represents a single feature associated with the data frame. So basically, there are 150 datapoints containing 4 features each.

`shape` here is an attribute of the dataframe and not a function, which is why it doesn't end in a pair of parentheses.

### `DataFrame.columns`
Let us now move into the 4 columns of data. What does each of them exactly represent? The`columns` attribute will give us the name of the columns in the dataframe.

In [3]:
iris_df.columns

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)'],
      dtype='str')

As we can see, there are four (4) columns. The `columns` attribute tells us the names of the columns and basically nothing else. This attribute assumes importance when we want to identify the features a dataset contains.

### `DataFrame.info`
The amount of data (given by the `shape` attribute) and the name of the features or columns (given by the `columns` attribute) tell us something about the dataset. Now, we would want to dive deeper into the dataset. The`DataFrame.info()` function is quite useful for this.

In [4]:
iris_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


From here, we can make a few observations:

1. The DataType of each column: In this dataset, all of the data is stored as 64-bit floating-point numbers.
2. Number of Non-Null values: Dealing with null values is an imortant step in data preparation. It will be dealt with later in the notebook

### DataFrame.describe()

Say we have a lot of numerical data in our dataset. Univariate statistical calculations such as the mean, median, quartiles etc. can be done on each of the columns individually. The `DataFrame.describe()` function provides us with a statistical summary of the numerical columns of a dataset.

In [5]:
iris_df.describe()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm)
count,150.000000,150.000000,150.000000,150.000000
mean,5.843333,3.057333,3.758000,1.199333
std,0.828066,0.435866,1.765298,0.762238
min,4.300000,2.000000,1.000000,0.100000
25%,5.100000,2.800000,1.600000,0.300000
50%,5.800000,3.000000,4.350000,1.300000
75%,6.400000,3.300000,5.100000,1.800000
max,7.900000,4.400000,6.900000,2.500000


The output above shows the total number of data points, mean, standard deviation, minimum, lower quartile(25%), median(50%), upper quartile(75%) and the maximum value of each column.

### `DataFrame.head()`
With all the above functions and attributes, we have got a top level view of the dataset. We know how many datapoints are there, how many features are there, the data type of each feature and the number of non-null values for each feature.

Now it's time to look at the data itself. Let's see what the first few rows (the first few datapoints) of our `DataFrame` look like:

In [6]:
iris_df.head()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm)
0,5.1,3.5,1.4,0.2
1,4.9,3.0,1.4,0.2
2,4.7,3.2,1.3,0.2
3,4.6,3.1,1.5,0.2
4,5.0,3.6,1.4,0.2


As the output here, we can see five (5) entries of the dataset. If we look at the index at the left, we find out that these are the first five rows.

### Exercise:
From the example given above, it is clear that, by default, `DataFrame.head()` returns the first five rows of a `DataFrame`. In the cell code below, can you figure out a way to display more than five rows?

In [7]:
# head(n) returns the first n rows of the DataFrame.
# So, to return for example 8 rows:
iris_df.head(8)

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm)
0,5.1,3.5,1.4,0.2
1,4.9,3.0,1.4,0.2
2,4.7,3.2,1.3,0.2
3,4.6,3.1,1.5,0.2
4,5.0,3.6,1.4,0.2
5,5.4,3.9,1.7,0.4
6,4.6,3.4,1.4,0.3
7,5.0,3.4,1.5,0.2


### `DataFrame.tail()`
Another way of looking at the data can be from the end (instead of the beginning). The flipside of `DataFrame.head()` is `DataFrame.tail()`, which returns the last five rows of a `DataFrame`:

In [8]:
iris_df.tail()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm)
145,6.7,3.0,5.2,2.3
146,6.3,2.5,5.0,1.9
147,6.5,3.0,5.2,2.0
148,6.2,3.4,5.4,2.3
149,5.9,3.0,5.1,1.8


In practice, it is useful to be able to easily examine the first few rows or the last few rows of a `DataFrame`, particularly when you are looking for outliers in ordered datasets.

All the functions and attributes shown above with the help of code examples, help us get a look and feel of the data.

> **Takeaway:** Even just by looking at the metadata about the information in a DataFrame or the first and last few values in one, you can get an immediate idea about the size, shape, and content of the data you are dealing with.

### Missing Data
Let us dive into missing data. Missing data occurs, when no value is stored in some of the columns.

Let us take an example: say someone is conscious about his/her weight and doesn't fill the weight field in a survey. Then, the weight value for that certain person will be missing.

Most of the time, in real world datasets, missing values occur.

**How Pandas Handles missing data**

Pandas handles missing values in two ways. The first you've seen before in previous sections: `NaN`, or Not a Number. This is actually a special value that is part of the IEEE floating-point specification and it is only used to indicate missing floating-point values.

For missing values apart from floats, pandas uses the Python `None` object. While it might seem confusing that you will encounter two different kinds of values that say essentially the same thing, there are sound programmatic reasong for this design choice and, in practice, going this route enables pandas to deliver a good compromise for the vast majority of cases. Notwithstanding this, both `None` and `NaN` carry restrictions that you need to be mindful of with regards to how they can be used.

### `None`: non-float missing data

Because `None` comes from Python, it cannot be used in NumPy and pandas arrays, that are not of data type `object`. Remember, NumPy arrays (and the data structures in pandas) can contain only one type of data. This is what gives them their tremendous power for large-scale data and computational work, but it also limits their flexibility. Such arrays have to upcast to the "lowest common denominator", the data type that will encompass everything in the array. When `None` is in the array, it means you are working with Python objects.

To see this in action, consider the following example array (note the `dtype` for it):

In [9]:
import numpy as np

example1 = np.array([2, None, 6, 8])
example1

array([2, None, 6, 8], dtype=object)

The reality of upcast data types carries two side effects with it. First, operations will be carried out at the level of interpreted Python code, rather than compiled NumPy code. Essentially, this means that any operations involving `Series` of `DataFrames` with `None` in them will be slower. While you would probably not notice this performance hit, for large datasets it might become an issue.

The second side effect stems from the first. Because `None` essentially drags `Series` or `DataFrames` back into the world of vanilla Python, using NumPy/pandas aggregations like `sum()` or `min()` on arrays that contain a `None` value will generally produce an error:

In [ ]:
example1.sum()

> **Key takeaway:** Addition (and other operations) between integers and `None` values is undefined, which can limit what you can do with datasets that contain them.

### `NaN`: missing float values
In contrast to `None`, NumPy (and therefore pandas) supports `NaN` for its fast vectorized operations and ufuncs. The bad news is that any arithmetic performed on `NaN`, always results in `NaN`. For example:

In [11]:
np.nan + 1

nan

In [12]:
np.nan * 0

nan

The good news: aggregations run on arrays with `NaN` in them dont't pop errors. The bad news: the results are not uniformly useful:

In [13]:
example2 = np.array([2, np.nan, 6, 8])
example2.sum(), example2.min(), example2.max()

(np.float64(nan), np.float64(nan), np.float64(nan))

### Exercise:

In [ ]:
# What happens if you add np.nan and None together?
np.nan + None

Remember: `NaN` is just for missing floating-point values; there is no `NaN` equivalent for integers, strings, or Booleans.

### `NaN` and `None`: null values in pandas
Even though `NaN` and `None` can behave somewhat differently, pandas is nevertheless built to handle them interchangeably. To see what we mean, consider a `Series` of integers:

In [15]:
int_series = pd.Series([1, 2, 3], dtype=int)
int_series

0    1
1    2
2    3
dtype: int64

### Exercise:

In [16]:
# Now set an element of int_series equal to None
int_series[0] = None

# How does that element shows up in the Series?
# a: NaN
int_series

# What is the dtype of the Series?
# a: float64
#    the initial int64 dtype, has been upcasted to float64

0    NaN
1    2.0
2    3.0
dtype: float64

In the process of upcasting data types to establish data homogeneity in `Series` and `DataFrames`, pandas will willingly switch missing values between `None` and `NaN`. Because of this design feature, it can be helpful to think of `None` and `NaN` as two different flavors of "null" in pandas. Indeed, some of the core methods you will use to deal with missing values in pandas reflect this idea in their names:

- **isnull()**  : Generates a Boolean mask indicating missing values
- **notnull()** : Opposite of isnull()
- **dropna()**  : Returns a filtered version of the data
- **fillna()**  : Returns a copy of the data with missing values filled or imputed

These are important methods to master and get comfortable with, so let's go over them each in some depth.

### Detecting null values

Now that we have understood the importance of missing values, we need to detect them in our dataset, before dealing with them. Both `isnull()` and `notnull()` are your primary methods for detecting null data. Both return Boolean masks over your data.

In [17]:
example3 = pd.Series([0, np.nan, "", None])

In [18]:
example3.isnull()

0    False
1     True
2    False
3     True
dtype: bool

Look closely at the output. Does any of it surprise you? While `0` is an artithmetic null, it's nevertheless a perfectly good integer and pandas treats it as such. "" is a little more subtle. While we used it in Section 1 to represent an empty string value, it is nevertheless a string object and not a representation of null, as far as pandas is concerned.

Now, let's turn this around and use these methods in a manner more like you will use them in practice. You can use Boolean masks directly as a `Series` or `DataFrame` index, which can be useful when trying to work with isolated missing (or present) values.

If we want the total number of missing values, we can just do a sum over the mask produced by the `isnull()` method.

In [19]:
example3.isnull().sum()

np.int64(2)

### Exercise:

In [20]:
# Try running example3[example3.notnull()]
# Before you do so, what do you expect to see?
example3.notnull()

0     True
1    False
2     True
3    False
dtype: bool

**Key takeaway**: Both the `isnull()` and `notnull()` methods produce similar results when you use them in DataFrames: they show the results and the index of those results, which will help you enormously as you wrestle with your data.

### Dealing with missing data



> **Learning goal:** By the end of this subsection, you should know how and when to replace or remove null values from DataFrames

Machine Learning models can't deal with missing data themselves. So, before passing the data into the model, we need to deal with these missing values.

How missing data is handled carries with it subtle tradeoffs, can affect your final analysis and real-world outcomes.

There are primarily two ways of dealing with missing data:

1. Drop the row containing the missing value
2. Replace the missing value with some other value

We will discuss both these methods and their pros and cons in details.

### Dropping null values

The amount of data we pass on to our model has a direct effect on its performance. Dropping null values means that we are reducing the number of datapoints, and hence reducing the size of the dataset. So, it is advisable to drop rows with null values when the dataset is quite large.

Another instance maybe that a certain row or column has a lot of missing values. Then, they maybe dropped because they wouldn't add much value to our analysis as most of the data is missing for that row/column.

Beyond identifying missing values, pandas provides a convenient means to remove null values from `Series` and `DataFrame`s. To see this in action, let's return to `example3`. The `DataFrame.dropna()` function helps in dropping the rows with null values.

In [21]:
example3 = example3.dropna()
example3

0    0
2     
dtype: object

Note that this should look like your output from `example3[example3.notnull()]`. The difference here is that, rather than just indexing on the masked values, `dropna()` has removed those missing values from the `Series example3`.

Because DataFrames have two dimensions, they afford more options for dropping data.

In [22]:
example4 = pd.DataFrame([[1, np.nan, 7], [2, 5, 8], [np.nan, 6, 9]])
example4

,0,1,2
0,1.0,NaN,7
1,2.0,5.0,8
2,NaN,6.0,9


(Did you notice that pandas upcast two of the columns to floats to accomodate the `NaN`s? The third columns remains int64 since there's no `NaN` in it)

You cannot drop a single value from a `DataFrame`, so you have to drop full rows or columns. Depending on what you are doing, you might want to do one or the other, and so pandas gives you options for both. Because in data science, column generally represent variables and rows represent observations, you are more likely to drop rows of data; the default setting for `dropna()` is to drop all rows that contain any null values:

In [23]:
example4.dropna()

,0,1,2
1,2.0,5.0,8


If for some reason you wish to drop NA values from columns, use `axis=1` attribute to do so:

In [24]:
example4.dropna(axis="columns")

,2
0,7
1,8
2,9


Notice that this can drop a lot of data that you might want to keep, paricularly in smaller datasets. What if you just want to drop rows or columns that contain several or even just all null values? You specify those settings in `dropna` with the `how` and `thresh` attributes.

By default, `how="any"` (if you would like to check for yourself or see what other parameters the method has, run`example4.dropn?` in a code cell). You could alternatively specify `how="all"` so as to drop only rows or columns that contain ALL null values. Let's expand our example `DataFrame` to see this in action in the next exercise.

In [25]:
example4[3] = np.nan
example4

,0,1,2,3
0,1.0,NaN,7,NaN
1,2.0,5.0,8,NaN
2,NaN,6.0,9,NaN


> Key takeaways:

1. Dropping null values is a good idea only if the dataset is large enough.
2. Full rows or columns can be dropped if they have most of their data missing.
3. The `DataFrame.dropna(axis=)` method helps in dropping null values. The `axis` argument signifies whether rows are to be dropped or columns.
4. The `how` argument can also be used. By default it is set to `any`. So, it drops only those rows/columns which contain any nymber of null values. It can be set to `all` to specify that we'll drop only those rows/columns, where all values are null.


### Exercise:

In [26]:
# How might you go about dropping just column 3?
# HINT: Remember that you will need to supply both the axis parameter and the how parameter.
example4.dropna(axis=1, how="all")

,0,1,2
0,1.0,NaN,7
1,2.0,5.0,8
2,NaN,6.0,9


The `thresh` parameter gives you finer-grained control: you set the number of *non-null* values that a row/column needs to have in order to be kept:

In [27]:
example4.dropna(axis="index", thresh=3)

,0,1,2,3
1,2.0,5.0,8,NaN


Here, the first and last row have been dropped, because they contain only two non-null values.

### Filling null values

It sometimes makes sense to fill in missing values with ones which could be valid. There are a few techiques to fill null values. The first is Domain Knowledge (knowledge of the subject on which the dataset is based) to somehow approximate the missing values.

You could use `isnull()` to do this inplace, but that can be laborious, paricularly if you have a lot of values to fill. Because this is such a common task in data science, pandas provides `fillna()` which returns a copy of the `Series` or `DataFrame` with the missing values replaced with one of your choosing. Let's create another example `Series` to see how this works in practice.

### Categorical Data (Non-numeric)

First let us consider non-numeric data. In datasets, we have columns with categorical data. e.g. Gender, True or False etc.

In most of these cases, we replace missing values with the `mode` of the column. Say, we have 100 data points and 90 have said True, 8 have said False and 2 have not filled. Then, we can will the 2 with True, considering the full column.

Again, here we can use domain knowledge here. Let us consider an example of filling with the `mode`.

In [28]:
fill_with_mode = pd.DataFrame([[1, 2, "True"], [3, 4, None], [5, 6, "False"], [7, 8, "True"], [9, 10, "True"]])
fill_with_mode

,0,1,2
0,1,2,True
1,3,4,NaN
2,5,6,False
3,7,8,True
4,9,10,True


Now, let's first find the mode before filling the `None` value with the mode.

In [29]:
fill_with_mode[2].value_counts()

2
True     3
False    1
Name: count, dtype: int64

So we will replace the missing value with True

In [30]:
fill_with_mode.fillna({2: "True"}, inplace=True)
fill_with_mode

,0,1,2
0,1,2,True
1,3,4,True
2,5,6,False
3,7,8,True
4,9,10,True


As we can see, the nul value has been replaced. Needless to say, we could have written anything in place of `True` and it would have got substituted.

### Numeric Data

Now, coming to numeric data. Here, we have two common ways of replacing missing values:

1. Replace with **Median** of the row
2. Replace with **Mean** of the row

We replace with Median, in case of skewed data with outliers. This is because median is robust to outliers.

When the data is normalized, we can use mean, as in that case, mean and median would be pretty close.

First, let us take a column which is normally distributed and let us fill the missing value with the mean of the column.

In [31]:
fill_with_mean = pd.DataFrame([[-2, 0, 1], [-1, 2, 3], [np.nan, 4, 5], [1, 6, 7], [2, 8, 9]])
fill_with_mean

,0,1,2
0,-2.0,0,1
1,-1.0,2,3
2,NaN,4,5
3,1.0,6,7
4,2.0,8,9


The mean of the column is zero:

In [32]:
# fill_with_mean[0].mean() or use
np.mean(fill_with_mean[0])

np.float64(0.0)

Filling with mean

In [33]:
fill_with_mean.fillna({0: np.mean(fill_with_mean[0])}, inplace=True)
fill_with_mean

,0,1,2
0,-2.0,0,1
1,-1.0,2,3
2,0.0,4,5
3,1.0,6,7
4,2.0,8,9


As we can see, the missing value has been replaced with the mean of the column.

Now let us try another dataframe, and this time we will replace the None values with the median of the column.

In [34]:
fill_with_median = pd.DataFrame(
    [
        [-2, 0, 1],
        [-1, 2, 3],
        [0, np.nan, 5],
        [1, 6, 7],
        [2, 8, 9],
    ]
)

fill_with_median

,0,1,2
0,-2,0.0,1
1,-1,2.0,3
2,0,NaN,5
3,1,6.0,7
4,2,8.0,9


The median of the second column is:

In [35]:
fill_with_median[1].median()

np.float64(4.0)

Filling with median:

In [36]:
fill_with_median.fillna({1: fill_with_median[1].median()}, inplace=True)
fill_with_median

,0,1,2
0,-2,0.0,1
1,-1,2.0,3
2,0,4.0,5
3,1,6.0,7
4,2,8.0,9


As we can see, the NaN value has been replaced by the median of the column.

In [37]:
example5 = pd.Series([1, np.nan, 2, None, 3], index=list("abcde"))
example5

a    1.0
b    NaN
c    2.0
d    NaN
e    3.0
dtype: float64

Fill all of the null entries with a single value, such as 0:

In [38]:
example5.fillna(0)

a    1.0
b    0.0
c    2.0
d    0.0
e    3.0
dtype: float64

> Key takeaways:

1. Filling in missing values should be done when either there is less data or there is a strategy to fill in the missing data.
2. Domain knowledge can be used to fill in missing values by approximating them.
3. For Categorical data, mostly, missing values are substituted with the **mode** of the column.
4. For numeric data, missing values are usually filled in with the mean (for normalized datasets) or the median of the columns.

### Exercise:

In [39]:
# What happens if you try to fill null values with a string, like ""?
example5.fillna("")

a    1.0
b       
c    2.0
d       
e    3.0
dtype: object

The missing values were replaced by a string but at the same time the dtype of the entries is now **object**!

You can **forward-fill** null values, which is to use the last valid value to fill a null:

In [40]:
example5.ffill()

a    1.0
b    1.0
c    2.0
d    2.0
e    3.0
dtype: float64

You can also **back-fill** to propagate the next valid value backward to fill a null:

In [41]:
example5.bfill()

a    1.0
b    2.0
c    2.0
d    3.0
e    3.0
dtype: float64

As you might guess, this works the same with DataFrames,  but you can also specify an `axis` along which to fill null values:

In [42]:
example4

,0,1,2,3
0,1.0,NaN,7,NaN
1,2.0,5.0,8,NaN
2,NaN,6.0,9,NaN


In [43]:
example4.ffill(axis=1)

,0,1,2,3
0,1.0,1.0,7.0,7.0
1,2.0,5.0,8.0,8.0
2,NaN,6.0,9.0,9.0


Notice that when a previous value is not available for forward-filling, the null value remains.

### Exercise:

In [44]:
# What output does example4.bfill(axis=1) produce?
# When there is no next valid value to fill, NaN remains
example4.bfill(axis=1)

,0,1,2,3
0,1.0,7.0,7.0,NaN
1,2.0,5.0,8.0,NaN
2,6.0,6.0,9.0,NaN


In [45]:
# A longer code snippet that can fill all of the null values in example4
# with the mean.
# NOTE: the default of mean() is to calculate the average over axis=0,
#       i.e. from top -> bottom
# So for the first column mean = (1+2)/2 = 1.5, and this value fills the
# last NaN cell of the column
example4.fillna(example4.mean())

,0,1,2,3
0,1.0,5.5,7,NaN
1,2.0,5.0,8,NaN
2,1.5,6.0,9,NaN


Notice that column 3 is still valueless: the default direction is to fill values row-wise.

> **Takeaway:** There are multiple ways to deal with missing values in your datasets. The specific strategy you use (removing them, replacing them, or even how you replace them) should be dictated by the particulars of that data. You will develop a better sense of how to deal with missing values the more you handle and interact with datasets.

### Encoding Categorical Data

Machine learnign models only deal with numbers and any form of numeric data. It won't be able to tell the difference between a Yes and a No, but it would be able to distinguish between 0 and 1. So, after filling in the missing values, we need to encode the categorical data to some numeric form for the model to understand.

Encoding can be done in two ways. We will be discussing them next.

**LABEL ENCODING**

Label encoding is basically converting each category to a number. For example, say we have a dataset of airline passengers and there is a column containing their class among the following ["business class", "economy class", "first class"]. If Label encoding is done on this, this would be transformed to [0, 1, 2]. Let us see an example via code. As we would be learning `scikit-learn` in the upcoming notebooks, we won't use it here.

In [46]:
label = pd.DataFrame(
    [
        [10, "business class"],
        [20, "first class"],
        [30, "economy class"],
        [40, "economy class"],
        [50, "economy class"],
        [60, "business class"],
    ],
    columns=["ID", "class"],
)

label

,ID,class
0,10,business class
1,20,first class
2,30,economy class
3,40,economy class
4,50,economy class
5,60,business class


To perform label encoding on the 1st column, we have to first describe a mapping from each class to a number, before replacing.

In [47]:
class_labels = {
    "business class": 0,
    "economy class": 1,
    "first class": 2,
}

label.replace({"class": class_labels}, inplace=True)

label

,ID,class
0,10,0
1,20,2
2,30,1
3,40,1
4,50,1
5,60,0


As we can see, the output matches what we thought would happen. So, when do we use label encoding? Label encoding is used in either or both of the following cases:

1. When the number of the categories is large
2. When the categories are in order.

**ONE HOT ENCODING**

Another type of encoding is One Hot Encoding. In this type of encoding, each category of the column gets added as a separate column and each datapoint will get a 0 or a 1 based on whether it contains that category. So, if there are n different categories, n columns will be appended to the dataframe.

For example, let us take the same aeroplane class example. The categories were:  ["business class", "economy class", "first class"]. So, if we perform one hot encoding, the following three columns  will be added to the dataset: ["class_business class", "class_economy class", "class_first class"].


In [48]:
one_hot = pd.DataFrame(
    [
        [10, "business class"],
        [20, "first class"],
        [30, "economy class"],
        [40, "economy class"],
        [50, "economy class"],
        [60, "business class"],
    ],
    columns=["ID", "class"],
)

one_hot

,ID,class
0,10,business class
1,20,first class
2,30,economy class
3,40,economy class
4,50,economy class
5,60,business class


Let us perform one hot encoding on the 1st column.

In [49]:
one_hot_data = pd.get_dummies(one_hot, columns=["class"])
one_hot_data

,ID,class_business class,class_economy class,class_first class
0,10,True,False,False
1,20,False,False,True
2,30,False,True,False
3,40,False,True,False
4,50,False,True,False
5,60,True,False,False


Each one hot encoded column contains 0 or 1, which specifies whether that category exists for that datapoint.

When do we use one hot encoding? One hot encoding is used in either or both of the following cases:

1. When the number of categories and the size of the dataset is smaller.
2. When the categories follow no particular order.

> Key Takeaways:

1. Encoding is done to convert non-numeric data to numeric data.
2. There are two types of encoding: Label encoding and One Hot encoding, both of which can be performed based on the demands of the dataset.

## Removing duplicate data

> **Learning goal:** By the end of this subsection, you should be comfortable identifying and removing duplicate values from DataFrames.

In addition to missing data, you will often encounter duplicated data in real-world datasets. Fortunately, pandas provides an easy means of detecting and removing duplicate entries.

### Identifying duplicates: `duplicated`

You can easily spot duplicate values using the `duplicated` method in pandas, which returns a Boolean mask indicating whether an entry in a `DataFrame` is a duplicate of an earlier one. Let's create another example `DataFrame` to see this in action.

In [50]:
example6 = pd.DataFrame(
    {
        "letters": ["A", "B"] * 2 + ["B"],
        "numbers": [1, 2, 1, 3, 3],
    }
)
example6

,letters,numbers
0,A,1
1,B,2
2,A,1
3,B,3
4,B,3


In [51]:
example6.duplicated()

0    False
1    False
2     True
3    False
4     True
dtype: bool

## Dropping duplicates: `drop_duplicates()`

`drop_duplicates()` simply returns a copy of the data for which all of the `duplicated` values are `False`:

In [52]:
example6.drop_duplicates()

,letters,numbers
0,A,1
1,B,2
3,B,3


Both `duplicated` and `drop_duplicates()` default to consider all columns but you can specify that they examine only a subset of columns in your `DataFrame`:

In [53]:
example6.drop_duplicates(["letters"])

,letters,numbers
0,A,1
1,B,2


> **Takeaway:** Removing duplicate data is an essential part of almost every data-science project. Duplicate data can change the results of your analyses and give you inaccurate results!

## Real-World Data Quality Checks

> **Learning goal:** By the end of this section, you should be comfortable detecting and correcting common real-world data quality issues including inconsistent categorical values, abnormal numerical values (outliers), and duplicate entities with variations.

While missing values and exact duplicates are common issues, real-world datasets often contain more subtle problems:

1. **Inconsistent categorical values**: The same category spelled differently (e.g. "USA", "U.S.A.", "United States")
2. **Abnormal numeric values**: Extreme outliers that indicate data entry errors (e.g. age = 999)
3. **Near-duplicate rows**: Records that represent the same entity with slight variations

Let's explore techniques to detect and handle these issues.

### Creating a Sample "Dirty" Dataset

First, let's create a sample dataset that contains the types of issues we commonly encounter in real-world data:

In [54]:
import pandas as pd
import numpy as np

# Create a sample dataset with quality issues
dirty_data = pd.DataFrame(
    {
        "customer_id": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12],
        "name": [
            "John Smith",
            "Jane Doe",
            "John Smith",
            "Bob Johnson",
            "Alice Williams",
            "Charlie Brown",
            "John  Smith",
            "Eva Martinez",
            "Bob Johnson",
            "Diana Prince",
            "Frank Castle",
            "Alice Williams",
        ],
        "age": [25, 32, 25, 45, 28, 199, 25, 31, 45, 27, -5, 28],
        "country": [
            "USA",
            "UK",
            "U.S.A",
            "Canada",
            "USA",
            "United Kingdom",
            "United States",
            "Mexico",
            "canada",
            "USA",
            "UK",
            "usa",
        ],
        "purchase_amount": [
            100.50,
            250.00,
            105.00,
            320.00,
            180.00,
            90.00,
            102.00,
            275.00,
            325.00,
            195.00,
            410.00,
            185.00,
        ],
    }
)

print("Sample 'Dirty' Dataset:")
print("=" * 70)
print(dirty_data)

Sample 'Dirty' Dataset:
    customer_id            name  age         country  purchase_amount
0             1      John Smith   25             USA            100.5
1             2        Jane Doe   32              UK            250.0
2             3      John Smith   25           U.S.A            105.0
3             4     Bob Johnson   45          Canada            320.0
4             5  Alice Williams   28             USA            180.0
5             6   Charlie Brown  199  United Kingdom             90.0
6             7     John  Smith   25   United States            102.0
7             8    Eva Martinez   31          Mexico            275.0
8             9     Bob Johnson   45          canada            325.0
9            10    Diana Prince   27             USA            195.0
10           11    Frank Castle   -5              UK            410.0
11           12  Alice Williams   28             usa            185.0


### 1. Detecting Inconsistent Categorical Values

Notice the `country` column has multiple representations for the same countries. Let's identify these inconsistencies:

In [55]:
# Check unique values in the country column
print("Unique Country Values:")
print("=" * 70)
print(dirty_data["country"].unique())

print(f"\nTotal Unique Values: {dirty_data["country"].nunique()}")
print("=" * 70)

# Count occurences of each unique variation
print("\nValue counts:")
print("=" * 70)
print(dirty_data["country"].value_counts())

Unique Country Values:
<StringArray>
[           'USA',             'UK',          'U.S.A',         'Canada',
 'United Kingdom',  'United States',         'Mexico',         'canada',
            'usa']
Length: 9, dtype: str

Total Unique Values: 9

Value counts:
country
USA               3
UK                2
U.S.A             1
Canada            1
United Kingdom    1
United States     1
Mexico            1
canada            1
usa               1
Name: count, dtype: int64


#### Standardizing Categorical Values

We can create a mapping to standardize these categorical values. A simple approach is to convert to lower case and create a mapping dictionary.

In [81]:
# Create a standardization mapping
country_mapping = {
    "usa": "USA",
    "u.s.a": "USA",
    "u.s.a.": "USA",
    "united states": "USA",
    "uk": "UK",
    "united kingdom": "UK",
    "canada": "Canada",
    "mexico": "Mexico",
}

# Standardize the country column
dirty_data["country_clean"] = dirty_data["country"].str.lower().map(country_mapping)

print("Before standardization:")
print("=" * 30)
print(dirty_data["country"].value_counts())
print("\nAfter standardization:")
print("=" * 30)
print(dirty_data["country_clean"].value_counts())

Before standardization:
country
USA               3
UK                2
U.S.A             1
Canada            1
United Kingdom    1
United States     1
Mexico            1
canada            1
usa               1
Name: count, dtype: int64

After standardization:
country_clean
USA       6
UK        3
Canada    2
Mexico    1
Name: count, dtype: int64


**Alternative: Using Fuzzy Matching**

For more complex cases, we can use fuzzy string matching with the `rapidfuzz` library to automatically detect similar strings:

In [ ]:
from rapidfuzz import process, fuzz

# Get unique countries
unique_countries = dirty_data["country"].unique()

# For each country find similar matches
print("Finding similar country names (similarity > 70%):")
print("=" * 70)

for country in unique_countries:
    matches = process.extract(country, unique_countries, scorer=fuzz.ratio, limit=3)

    # Filter matches with similarity > 70 and not identical.
    similar = [m for m in matches if m[1] > 70 and m[0] != country]
    if similar:
        print(f"\n'{country}' is similar to :")
        for match, score, _ in similar:
            print(f"  - '{match}' (similarity: {score}%)")

Finding similar country names (similarity > 70%):

'USA' is similar to :
  - 'U.S.A' (similarity: 75.0%)

'U.S.A' is similar to :
  - 'USA' (similarity: 75.0%)

'Canada' is similar to :
  - 'canada' (similarity: 83.33333333333334%)

'canada' is similar to :
  - 'Canada' (similarity: 83.33333333333334%)


### 2. Detecting Abnormal Numeric Values (Outliers)

Looking at the `age` column, we have some suspicious values like 199 and -5. Let's use statistical methods to detect these outliers.

In [61]:
# Display basic statistis
print("Age column statistics:")
print("=" * 30)
print(dirty_data["age"].describe())

# Identify impossible values using domain knowledge
print("\nRows with impossible age values (age < 0 or age > 120):")
print("=" * 60)
impossible_ages = dirty_data[(dirty_data["age"] < 0) | (dirty_data["age"] > 120)]
print(impossible_ages[["customer_id", "name", "age"]])

Age column statistics:
count     12.000000
mean      42.083333
std       50.973181
min       -5.000000
25%       25.000000
50%       28.000000
75%       35.250000
max      199.000000
Name: age, dtype: float64

Rows with impossible age values (age < 0 or age > 120):
    customer_id           name  age
5             6  Charlie Brown  199
10           11   Frank Castle   -5


#### Using IQR (Interquartile Range) Method

The IQR method is a robust statistical technique for outlier detection that is less sensitive to extreme values:

In [62]:
# Calculate IQR for age (excluding impossible values)
valid_ages = dirty_data[(dirty_data["age"] >= 0) | (dirty_data["age"] < 120)]["age"]

Q1 = valid_ages.quantile(0.25)
Q3 = valid_ages.quantile(0.75)
IQR = Q3 - Q1

# Define outlier bounds
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print(f"IQR-based outlier bounds for age: [{lower_bound:.2f}, {upper_bound:.2f}]")

# Identify outliers
age_outliers = dirty_data[(dirty_data["age"] < lower_bound) | (dirty_data["age"] > upper_bound)]
print(f"\nRows with age outliers:")
print("=" * 40)
print(age_outliers[["customer_id", "name", "age"]])

IQR-based outlier bounds for age: [9.62, 50.62]

Rows with age outliers:
    customer_id           name  age
5             6  Charlie Brown  199
10           11   Frank Castle   -5


#### Using Z-Score Method

The Z-Score method identifies outliers based on standard deviations from the mean:

In [70]:
from scipy import stats

# Calculate Z-scores for age, handling NaN values
age_nonan = dirty_data["age"].dropna()
zscores = np.abs(stats.zscore(age_nonan))
dirty_data["age_zscore"] = np.nan
dirty_data.loc[age_nonan.index, "age_zscore"] = zscores

# Typically, Z-score > 3 indicates an outlier
print("Rows with age Z-score > 3:")
print("=" * 50)
zscore_outliers = dirty_data[dirty_data["age_zscore"] > 3]
print(zscore_outliers[["customer_id", "name", "age", "age_zscore"]])

# Clean up the temp column
dirty_data = dirty_data.drop("age_zscore", axis=1)

Rows with age Z-score > 3:
   customer_id           name  age  age_zscore
5            6  Charlie Brown  199    3.215301


#### Handling Outliers

Once detected, outliers can be handled in several ways:

1. **Remove**: Drop rows with outliers if they're errors
2. **Cap**: Replace with boundary values
3. **Replace with NaN**: Treat as missing data and use imputation techniques
4. **Keep**: If they're legitimate extreme values

In [71]:
# Create a cleaned version by replacing impossible ages with NaN
dirty_data["age_clean"] = dirty_data["age"].apply(lambda x: np.nan if (x < 0 or x > 120) else x)

print("Age column before and after cleaning:")
print("=" * 50)
print(dirty_data[["customer_id", "name", "age", "age_clean"]])

Age column before and after cleaning:
    customer_id            name  age  age_clean
0             1      John Smith   25       25.0
1             2        Jane Doe   32       32.0
2             3      John Smith   25       25.0
3             4     Bob Johnson   45       45.0
4             5  Alice Williams   28       28.0
5             6   Charlie Brown  199        NaN
6             7     John  Smith   25       25.0
7             8    Eva Martinez   31       31.0
8             9     Bob Johnson   45       45.0
9            10    Diana Prince   27       27.0
10           11    Frank Castle   -5        NaN
11           12  Alice Williams   28       28.0


### 3. Detecting Near-Duplicate Rows

Notice that our dataset has multiple entries for "John Smith" with slightly different values. Let's identify potential duplicates based on name similarity.

In [72]:
# First, let's look at exact name matches (ignoring extra whitespace)
dirty_data["name_normalized"] = dirty_data["name"].str.strip().str.lower()

print("Checking for duplicate names:")
print("=" * 50)

# mark all duplicates as True
duplicate_names = dirty_data[dirty_data.duplicated(["name_normalized"], keep=False)]
print(duplicate_names.sort_values("name_normalized")[["customer_id", "name", "age", "country"]])

Checking for duplicate names:
    customer_id            name  age country
4             5  Alice Williams   28     USA
11           12  Alice Williams   28     usa
3             4     Bob Johnson   45  Canada
8             9     Bob Johnson   45  canada
0             1      John Smith   25     USA
2             3      John Smith   25   U.S.A


Notice that row: 6 "John  Smith" is NOT included in the duplicates. It contains two whitespace characters between the names, so it's considered as a different name!

#### Finding Near-Duplicates with Fuzzy Matching

For more sophisticated duplicate detection, we can use fuzzy matching to find similar names:

In [75]:
from rapidfuzz import process, fuzz


# Function to find potential duplicates
def find_near_duplicates(df: pd.DataFrame, column: str, threshold: int = 90):
    """
    Find near duplicate entries in a column using fuzzy matching.

    Parameters:
    - df: DataFrame
    - column: Column name to check for duplicates
    - threshold: Similarity threshold (0-100)

    Returns: List of potential duplicate groups
    """
    values = df[column].unique()
    duplicate_groups = []
    checked = set()

    for value in values:
        if value in checked:
            continue

        # Find similar values
        matches = process.extract(value, values, scorer=fuzz.ratio, limit=len(values))
        similar = [m[0] for m in matches if m[1] >= threshold]

        if len(similar) > 1:
            duplicate_groups.append(similar)
            checked.update(similar)

    return duplicate_groups


# Find new-duplicate names
duplicate_groups = find_near_duplicates(dirty_data, "name", threshold=90)

print("Potential duplicate groups:")
print("=" * 50)

for i, group in enumerate(duplicate_groups, 1):
    print(f"\nGroup {i}:")
    for name in group:
        matching_rows = dirty_data[dirty_data["name"] == name]
        print(f"  '{name}': {len(matching_rows)} occurence(s)")
        for _, row in matching_rows.iterrows():
            print(f"   - Customer {row['customer_id']}: age={row['age']}, country={row['country']}")

Potential duplicate groups:

Group 1:
  'John Smith': 2 occurence(s)
   - Customer 1: age=25, country=USA
   - Customer 3: age=25, country=U.S.A
  'John  Smith': 1 occurence(s)
   - Customer 7: age=25, country=United States


#### Handling Duplicates

Once identified, you need to decide how to handle duplicates:

1. **Keep the first occurence**: Use `drop_duplicates(keep="first")`
2. **Keep the last occurence**: Use `drop_duplicates(keep="last")
3. **Aggregate information**: Combine information from duplicate rows
4. **Manual review**: Flag for human review

In [76]:
# Example: Remove duplicates based on normalized name, keeping first occurence
cleaned_data = dirty_data.drop_duplicates(keep="first", subset=["name_normalized"])

print(f"Original dataset: {len(dirty_data)} rows")
print(f"After removing name duplicates: {len(cleaned_data)} rows")
print(f"Removed: {len(dirty_data)-len(cleaned_data)} duplicate rows")

print("\nCleaned dataset:")
print("=" * 50)
print(cleaned_data[["customer_id", "name", "age", "country_clean"]])

Original dataset: 12 rows
After removing name duplicates: 9 rows
Removed: 3 duplicate rows

Cleaned dataset:
    customer_id            name  age country_clean
0             1      John Smith   25           USA
1             2        Jane Doe   32            UK
3             4     Bob Johnson   45        Canada
4             5  Alice Williams   28           USA
5             6   Charlie Brown  199            UK
6             7     John  Smith   25           USA
7             8    Eva Martinez   31        Mexico
9            10    Diana Prince   27           USA
10           11    Frank Castle   -5            UK


### Summary: Complete Data Cleaning Pipeline

Let's put it all together into a comprehensive cleaning pipeline:

In [77]:
def clean_dataset(df: pd.DataFrame) -> pd.DataFrame:
    """Comprehensive data cleaning function."""
    # Create a copy to avoid modifying the original
    cleaned = df.copy()

    # 1. Standardize categorical values (country)
    country_mapping = {
        "usa": "USA",
        "u.s.a": "USA",
        "united states": "USA",
        "uk": "UK",
        "united kingdom": "UK",
        "canada": "Canada",
        "mexico": "Mexico",
    }
    cleaned["country"] = cleaned["country"].str.lower().map(country_mapping)

    # 2. Clean abnormal age values
    cleaned["age"] = cleaned["age"].apply(lambda x: np.nan if (x < 0 or x > 120) else x)

    # 3. Remove near-duplicate names (normalize whitespace)
    cleaned["name"] = cleaned["name"].str.strip()
    cleaned = cleaned.drop_duplicates(subset=["name"], keep="first")

    return cleaned


# Apply the cleaning pipeline
final_cleaned_data = clean_dataset(dirty_data)

print("Before cleaning:")
print(f"  Rows: {len(dirty_data)}")
print(f"  Unique countries: {dirty_data["country"].nunique()}")
print(f"  Invalid ages: {((dirty_data["age"]<0) | (dirty_data["age"]>120)).sum()}")

print("\nAfter cleaning:")
print(f"  Rows: {len(final_cleaned_data)}")
print(f"  Unique countries: {final_cleaned_data["country"].nunique()}")
print(f"  Invalid ages: {((final_cleaned_data["age"]<0) | (final_cleaned_data["age"]>120)).sum()}")

print("\nCleaned dataset:")
print(final_cleaned_data[["customer_id", "name", "age", "country", "purchase_amount"]])

Before cleaning:
  Rows: 12
  Unique countries: 9
  Invalid ages: 2

After cleaning:
  Rows: 9
  Unique countries: 4
  Invalid ages: 0

Cleaned dataset:
    customer_id            name   age country  purchase_amount
0             1      John Smith  25.0     USA            100.5
1             2        Jane Doe  32.0      UK            250.0
3             4     Bob Johnson  45.0  Canada            320.0
4             5  Alice Williams  28.0     USA            180.0
5             6   Charlie Brown   NaN      UK             90.0
6             7     John  Smith  25.0     USA            102.0
7             8    Eva Martinez  31.0  Mexico            275.0
9            10    Diana Prince  27.0     USA            195.0
10           11    Frank Castle   NaN      UK            410.0


### 🎯 Challenge Exercise

Now it's your turn! Below is a new row of data with multiple quality issues. Can you:

1. Identify all the issues in this row
2. Write code to clean each issue
3. Add the cleaned row to the dataset

Here's the problematic data:

In [82]:
# New problematic row:
new_row = pd.DataFrame(
    {
        "customer_id": [13],
        "name": ["  Diana Prince  "],  # Extra whitespce
        "age": [250],  # Impossible age
        "country": ["U.S.A."],  # Insonsistent format
        "purchase_amount": [150.00],
    }
)

print("New row to clean:")
print(new_row)

# TODO: Your code here to clean this row
#
# 1. Strip whitespace from the name
new_row_cleaned = new_row.copy()
new_row_cleaned["name"] = new_row_cleaned["name"].str.strip()

# 2. Check if the name is a duplicate (Dian Prince already exists)


# 3. Handle the impossible age value
new_row_cleaned["age"] = new_row_cleaned["age"].apply(lambda x: np.nan if (x < 0 or x > 120) else x)

# 4. Standardize the country name
new_row_cleaned["country"] = new_row_cleaned["country"].str.lower().map(country_mapping)

print("\nCleaned row:")
print(new_row_cleaned)

New row to clean:
   customer_id              name  age country  purchase_amount
0           13    Diana Prince    250  U.S.A.            150.0

Cleaned row:
   customer_id          name  age country  purchase_amount
0           13  Diana Prince  NaN     USA            150.0


### Key Takeaways

1. **Inconsistent categories** are common in real-world data. Always check unique values and standardize them using mappings or fuzzy matching.

2. **Outliers** can significantly affect your analysis. Use domain knowledge combined with statistical methods (IQR, Z-score) to detect them.

3. **Near-duplicates** are harder to detect than exact duplicates. Consider using fuzzy matching and normalizing data (lowercasing, stripping whitespace) to identify them.

4. **Data cleaning is iterative**. You may need to apply multiple techniques and review the results before finalizing your cleaned dataset.

5. **Document your decisions**. Keep track of what cleaning steps you applied and why, as this important for reproducibility and transparency.

> **Best Practice:** Always keep a copy of your original "dirty" data. Never overwrite your source data files - create cleaned versions with clear naming conventions like `data_cleaned.csv`.